[Model Element] Enforce a default memory limit on entity loads

achan00 opened the pull request:
#### c0d5482e12f00b4dfb45eeadafe5ae738891cdc0
<pre>
[Model Element] Enforce a default memory limit on entity loads
<a href="https://bugs.webkit.org/show_bug.cgi?id=295167">https://bugs.webkit.org/show_bug.cgi?id=295167</a>
<a href="https://rdar.apple.com/145610624">rdar://145610624</a>

Reviewed by Mike Wyrzykowski.

Specify enforceMemoryConstraints and memoryLimit in load options when loading
the model entities. The default memory limit that ModelProcessModelPlayerProxy
uses is 100MB. For testing/debugging purposes, this memory limit can be
overridden with the WebKitDebugModelEntityMemoryLimit runtime default.

* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp:
(WebKit::ModelConnectionToWebProcess::debugEntityMemoryLimit const):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
* Source/WebKit/ModelProcess/ModelProcess.cpp:
(WebKit::ModelProcess::initializeModelProcess):
Store the entity memory limit from the model process creation parameters.
* Source/WebKit/ModelProcess/ModelProcess.h:
* Source/WebKit/ModelProcess/ModelProcessCreationParameters.h:
* Source/WebKit/ModelProcess/ModelProcessCreationParameters.serialization.in:
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.cpp:
(WebKit::ModelProcessModelPlayerManagerProxy::createModelPlayer):
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::RKModelLoaderUSD::load):
Pass the entity memory limit to -[WKRKEntity loadFromData:...]. If no
memory limit is specified, pass in 0 to tell WKRKEntity to load without limit.
(WebKit::loadREModelUsingRKUSDLoader):
(WebKit::ModelProcessModelPlayerProxy::create):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::load):
Pass the entity memory limit to loadREModelUsingRKUSDLoader(). If no memory
limit is set, pass in the default which is 100MB.
* Source/WebKit/UIProcess/Model/cocoa/ModelProcessProxyCocoa.mm:
(WebKit::ModelProcessProxy::updateModelProcessCreationParameters):
* Source/WebKit/WebKitSwift/RealityKit/WKRKEntity.h:
* Source/WebKit/WebKitSwift/RealityKit/WKRKEntity.swift:
(WKRKEntity.load(from:withAttributionTaskID:entityMemoryLimit:)):
(WKRKEntity.load(from:withAttributionTaskID:)): Deleted.

Canonical link: <a href="https://commits.webkit.org/296796@main">https://commits.webkit.org/296796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5eea27c10851c74880246ad34052b1810a05eeb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109591 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19678 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115643 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59856 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37837 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83298 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112539 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98700 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63757 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23216 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59407 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93212 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16883 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118436 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36630 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27120 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92306 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37003 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94960 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92127 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23471 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37060 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14807 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32440 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36524 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41995 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36184 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39527 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37894 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/WebKitSwift/RealityKit/WKRKEntity.swift:
```diff
@@ -56,14 +56,18 @@ extension WKRKEntity {
             }
+            if entityMemoryLimit > 0 {
+                loadOptions.enforceMemoryConstraints = true
+                loadOptions.memoryLimit = entityMemoryLimit * 1024 * 1024
```
`wtf/StdLibExtras.h` defines `MB` as:
```
static constexpr size_t MB = 1024 * 1024;
```
could use that in place, but in any case it's fairly obvious

achan00 replied:
I'm not sure how to reference `WTF::MB` in the Swift file. @rr-codes Is that possible?

rr-codes replied:
it isn't; you can use the https://developer.apple.com/documentation/foundation/measurement type though

achan00 replied:
Thanks! `Measurement` should work but it would actually require a couple of Double <-> Int conversions as `Measurement` works with `Double` but the memory limit is expecting `Int`. Not sure if this is worth the extra complexity.

mwyrzykowski replied:
I'd say either leave as-is or alternatively you can pass `entityMemoryLimit` as bytes instead of megabytes but not necessary

mwyrzykowski commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -609,6 +612,8 @@ static RESRT modelLocalizedTransformSRT(RESRT originalSRT)
 
 // MARK: - WebCore::ModelPlayer
 
+static int defaultEntityMemoryLimit = 100; // MB
```
100 MB seems a bit low for complex models but perhaps this is fine? Simple models could likely be <1 MB but more complex ones could be very large, perhaps we are not interested in supporting those use cases right now?

achan00 replied:
We are imposing a limit so the model process' memory usage can be under some control. RE will try to downsample to fit that limit if possible.

achan00 committed:
[Model Element] Enforce a default memory limit on entity loads
https://bugs.webkit.org/show_bug.cgi?id=295167
rdar://145610624

Reviewed by Mike Wyrzykowski.

Specify enforceMemoryConstraints and memoryLimit in load options when loading
the model entities. The default memory limit that ModelProcessModelPlayerProxy
uses is 100MB. For testing/debugging purposes, this memory limit can be
overridden with the WebKitDebugModelEntityMemoryLimit runtime default.

* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp:
(WebKit::ModelConnectionToWebProcess::debugEntityMemoryLimit const):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
* Source/WebKit/ModelProcess/ModelProcess.cpp:
(WebKit::ModelProcess::initializeModelProcess):
Store the entity memory limit from the model process creation parameters.
* Source/WebKit/ModelProcess/ModelProcess.h:
* Source/WebKit/ModelProcess/ModelProcessCreationParameters.h:
* Source/WebKit/ModelProcess/ModelProcessCreationParameters.serialization.in:
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.cpp:
(WebKit::ModelProcessModelPlayerManagerProxy::createModelPlayer):
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::RKModelLoaderUSD::load):
Pass the entity memory limit to -[WKRKEntity loadFromData:...]. If no
memory limit is specified, pass in 0 to tell WKRKEntity to load without limit.
(WebKit::loadREModelUsingRKUSDLoader):
(WebKit::ModelProcessModelPlayerProxy::create):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::load):
Pass the entity memory limit to loadREModelUsingRKUSDLoader(). If no memory
limit is set, pass in the default which is 100MB.
* Source/WebKit/UIProcess/Model/cocoa/ModelProcessProxyCocoa.mm:
(WebKit::ModelProcessProxy::updateModelProcessCreationParameters):
* Source/WebKit/WebKitSwift/RealityKit/WKRKEntity.h:
* Source/WebKit/WebKitSwift/RealityKit/WKRKEntity.swift:
(WKRKEntity.load(from:withAttributionTaskID:entityMemoryLimit:)):
(WKRKEntity.load(from:withAttributionTaskID:)): Deleted.

Canonical link: https://commits.webkit.org/296796@main
